util/oteltest/src/oteltest/__init__.py

```diff
@@ -0,0 +1,172 @@
+
+
+import abc
+import dataclasses
+import json
+from typing import List, Mapping, Optional, Sequence
+
+
+class OtelTest(abc.ABC):
+    """
+    Abstract base class for tests using OtelTest. No need to instantiate or call it -- just define a subclass of
+    OtelTest anywhere in your script.
+
+    The first three methods are for configuration and the second two are callbacks.
+
+    When you run the `oteltest` command against your script, a new Python virtual environment is created with the
+    configuration specified by this class's implementation. The two callbacks are then run, but not in the subprocess
+    of the script, rather in the process of the oteltest command.
+    """
+
+    @abc.abstractmethod
+    def environment_variables(self) -> Mapping[str, str]:
+        """
+        Return a mapping of environment variables to their values. These will become the env vars for your script.
+        """
+
+    @abc.abstractmethod
+    def requirements(self) -> Sequence[str]:
+        """
+        Return a sequence of requirements for the script. Each string should be formatted as for `pip install`.
+        """
+
+    @abc.abstractmethod
+    def wrapper_command(self) -> str:
+        """
+        Return a wrapper command to run the script. For example, return 'opentelemetry-instrument' for
+        `opentelemetry-instrument python my_script.py`. Return an empty string for`python my_script.py`.
+        """
+
+    @abc.abstractmethod
+    def on_start(self) -> Optional[float]:
+        """
+        Called immediately after the script has started.
+
+        You can add client code here to call your server and have it produce telemetry.
+        Note: you may need to sleep or check for liveness first!
+        Note: this method will run in the Python virtual environment of `oteltest`, not that of the script.
+
+        Return a float indicating how long to wait in seconds for the script to finish. Once that time has elapsed,
+        the subprocess running the script will be force terminated. Or return `None` to allow the script to finish on
+        its own, in which case the script should shut itself down.
+        """
+
+    @abc.abstractmethod
+    def on_stop(
+        self, telemetry: "Telemetry", stdout: str, stderr: str, returncode: int
+    ) -> None:
+        """
+        Called immediately after the script has ended. Passed in are both the telemetry otelsink received while the
+        script was running and the output of the script (stdout, stderr, returncode).
+        """
+
+
+@dataclasses.dataclass
+class Request:
+    request: dict
+    headers: dict
+    test_elapsed_ms: int
+
+    def get_header(self, name):
+        return self.headers.get(name)
+
+    def to_json(self):
+        return json.dumps(self.to_dict())
+
+    def to_dict(self):
+        return {
+            "request": self.request,
+            "headers": self.headers,
+            "test_elapsed_ms": self.test_elapsed_ms,
+        }
+
+
+class Telemetry:
```

Comment:
A docstring might help here too. It does feel obvious to me what it's for but maybe not for newer users.

Reply:
Good idea -- done.

Reply:
Awesome